src/_pytest/nodes.py

```diff
@@ -93,26 +93,40 @@ def iterparentnodeids(nodeid: str) -> Iterator[str]:
         yield nodeid
 
 
+def _check_path(path: Path, fspath: LEGACY_PATH) -> None:
+    if Path(fspath) != path:
+        raise ValueError(
+            f"Path({fspath!r}) != {path!r}\n"
+            "if both path and fspath are given they need to be equal"
+        )
+
+
 def _imply_path(
     path: Optional[Path], fspath: Optional[LEGACY_PATH]
 ) -> Tuple[Path, LEGACY_PATH]:
     if path is not None:
         if fspath is not None:
-            if Path(fspath) != path:
-                raise ValueError(
-                    f"Path({fspath!r}) != {path!r}\n"
-                    "if both path and fspath are given they need to be equal"
-                )
-            assert Path(fspath) == path, f"{fspath} != {path}"
+            _check_path(path, fspath)
         else:
             fspath = legacy_path(path)
         return path, fspath
-
     else:
         assert fspath is not None
         return Path(fspath), fspath
 
 
+# Optimization: use _imply_path_only over _imply_path when only need Path.
+#               This is to avoid `legacy_path(path)` which is suprisngly heavy.
```

Comment:
```suggestion
#               This is to avoid `legacy_path(path)` which is surprisingly heavy.
```

Reply:
Fixed, thanks. I also added a bit to a test which would hopefully fix the missing coverage.